pkgs/top-level/java-packages.nix

```diff
@@ -7,27 +7,46 @@ with pkgs;
 
   compiler =
     let
+      mergeMetaPlatforms =
+        jdk: other:
+        jdk
+        // {
+          meta = jdk.meta // {
+            platforms = lib.unique (jdk.meta.platforms ++ other.meta.platforms);
+          };
+        };
+
       mkOpenjdk =
         featureVersion: path-darwin:
-        if stdenv.hostPlatform.isLinux then
-          mkOpenjdkLinuxOnly featureVersion
-        else
-          let
-            openjdk = callPackage path-darwin { };
-          in
-          openjdk // { headless = openjdk; };
-
-      mkOpenjdkLinuxOnly =
-        featureVersion:
         let
-          openjdk = callPackage ../development/compilers/openjdk/generic.nix { inherit featureVersion; };
+          # merge meta.platforms of both packages so that dependent packages and hydra build them
+          mergeMetaPlatforms =
+            jdk: other:
+            jdk
+            // {
+              meta = jdk.meta // {
+                platforms = lib.unique (jdk.meta.platforms ++ other.meta.platforms);
+              };
+            };
+          openjdkLinux =
+            let
+              openjdk = callPackage ../development/compilers/openjdk/generic.nix { inherit featureVersion; };
+            in
+            openjdk
+            // {
+              headless = mergeMetaPlatforms openjdkLinuxHeadless openjdkDarwin;
+            };
+          openjdkLinuxHeadless = openjdkLinux.override { headless = true; };
+          openjdkDarwin =
+            let
+              openjdk = callPackage path-darwin { };
+            in
+            openjdk // { headless = mergeMetaPlatforms openjdkDarwin openjdkLinuxHeadless; };
         in
-        assert stdenv.hostPlatform.isLinux;
-        openjdk
-        // {
-          headless = openjdk.override { headless = true; };
-        };
-
+        if stdenv.hostPlatform.isLinux then
+          (mergeMetaPlatforms openjdkLinux openjdkDarwin)
+        else
+          (mergeMetaPlatforms openjdkDarwin openjdkLinux);
```

Comment:
I feel like this specific pattern could be pulled out into a function, since it's repeated 3 times. Not sure how much it would affect readability though.